### PR TITLE
Layered em

### PR DESCRIPTION
**Pull request by jcapriot**
This adds the analytic function of a vertical magnetic dipole in a half space from Ward and Hohmann's Chapter.

It also adds the semi-analytic form of the layered half-space solution. This is done analytically in the wavenumber domain, and then transformed back to the space domain using the filters from empymod. The kernel function also has the necessary derivatives implemented for calculating Jacobins with respect to these operations.

There are two versions of the kernel functions (which actually both run in similar time on my machine at this point).
The first is a pure numpy solution. The second is an optional cython+cpp implementation. These algorithms both do the same thing (and are tested to match). The largest difference will be that the numpy version will require more memory to store the intermediate arrays necessary for broadcasting.

To enable building with the C extensions, the setup.py files will conditionally add the cython files to the build if the environment variable BUILD_GEOANA_EXT is set to something besides 0. This should allow us to still distribute the source code just fine for people who do not want to compile anything on the backend.

I do still need to double check that the cython source files will still be part of the source distribution that is uploaded however...

### CODE REVIEW

File: geoana/kernels/tranverse_electric_reflections.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+    Parameters
+    ----------
+    frequencies : float, numpy.ndarray
+        Frequency (Hz); shape = (n_frequency, ).
+    lamb : float, numpy.ndarray
+        Spatial wavenumber (1/m), shape = (n_filter, ).
+    sigma: complex, numpy.ndarray
+        Conductivity (S/m), shape = (n_layer, n_frequency).
+    chi: complex, numpy.ndarray
```
Comment (lheagy) [inline geoana/kernels/tranverse_electric_reflections.py:18]: do we want to work with `chi` or `mu` here? I would be tempted to stick with `mu` throughout for electromagnetic problems

File: geoana/kernels/tranverse_electric_reflections.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
+import numpy as np
+from scipy.constants import mu_0
+
+def _rTE_forward(frequencies, lamb, sigma, chi, thicknesses):
+    """Compute reflection coefficients for Transverse Electric (TE) mode.
+
+    The first layer is considered to be the top most layer. The last
+    layer is considered to have infinite thickness.
```
Comment (lheagy) [inline geoana/kernels/tranverse_electric_reflections.py:8]:
```suggestion
    layer is considered to have infinite thickness. All physical properties 
    are defined starting from the top most layer. 
```

File: geoana/em/fdem/layered.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+from scipy.constants import mu_0, epsilon_0
+from empymod.utils import check_hankel
+from empymod.transform import get_dlf_points
+from geoana.kernels.tranverse_electric_reflections import rTE_forward
+
+
+class MagneticDipoleLayeredHalfSpace(BaseMagneticDipole, BaseFDEM):
+
+    thickness = properties.Array(
+        "Layer thicknesses (m)",
```
Comment (lheagy) [inline geoana/em/fdem/layered.py:14]:
```suggestion
        "Layer thicknesses (m) starting from the top-most layer. The bottom layer is assumed to be infinite.",
```

File: geoana/em/fdem/layered.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        shape=('*', ),
+        dtype=float
+    )
+
+    sigma = properties.Array(
+        "Electrical conductivity (S/m)",
```
Comment (lheagy) [inline geoana/em/fdem/layered.py:20]:
```suggestion
        "Electrical conductivity (S/m), defined starting from the top most layer",
```

**Review by lheagy — APPROVED** [review]
This looks good, thanks @jcapriot! My only real comment is that I would suggest we stick with `mu` being the default magnetic property that we pass around in the code (at least for EM)

**Comment by jcapriot** [comment]
yeah, I'm happy with that too, should be easy to change quickly

File: geoana/em/fdem/layered.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    )
+
+    frequency = properties.Array(
+        "Source frequency (Hz)",
+        shape=('*', ),
+        dtype=float
+    )
+
+    mu = properties.Array(
+        "Magnetic permeability (H/m)",
```
Comment (jcapriot) [inline geoana/em/fdem/layered.py:33]:
```suggestion
        "Magnetic permeability (H/m), defined starting from the top most layer",
```

File: geoana/em/fdem/layered.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        shape=('*', ),
+        dtype=complex,
+        default=np.array([mu_0], dtype=np.complex128)
+    )
+
+    epsilon = properties.Array(
+        "Permitivity value (F/m)",
```
Comment (jcapriot) [inline geoana/em/fdem/layered.py:40]:
```suggestion
        "Permitivity value (F/m), defined starting from the top most layer",
```

File: geoana/kernels/_extensions/rTE.pyx [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+        SIZE_t n_frequency,
+        SIZE_t n_filter,
+        SIZE_t n_layers
+    ) nogil
+
+def rTE_forward(frequencies, lamb, sigma, mu, thicknesses):
+    """Compute reflection coefficients for Transverse Electric (TE) mode.
+
+    The first layer is considered to be the top most layer. The last
+    layer is considered to have infinite thickness.
```
Comment (jcapriot) [inline geoana/kernels/_extensions/rTE.pyx:45]:
```suggestion
    layer is considered to have infinite thickness. All physical properties 
    are defined starting from the top most layer. 
```

File: geoana/kernels/_extensions/rTE.pyx [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    rTE(out_p, &f[0], &lam[0], sig_p, &c_mu[0, 0], h_p,
+          n_frequency, n_filter, n_layers)
+
+    return np.array(out)
+
+def rTE_gradient(frequencies, lamb, sigma, mu, thicknesses):
+    """Compute reflection coefficients for Transverse Electric (TE) mode.
+
+    The first layer is considered to be the top most layer. The last
+    layer is considered to have infinite thickness.
```
Comment (jcapriot) [inline geoana/kernels/_extensions/rTE.pyx:122]:
```suggestion
    layer is considered to have infinite thickness. All physical properties 
    are defined starting from the top most layer. 
```

File: geoana/kernels/tranverse_electric_reflections.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    Y0 = lamb/(1j*omega*mu_0)[:, None]
+    TE = (Y0 - Yh)/(Y0 + Yh)
+
+    return TE
+
+def _rTE_gradient(frequencies, lamb, sigma, mu, thicknesses):
+    """Compute reflection coefficients for Transverse Electric (TE) mode.
+
+    The first layer is considered to be the top most layer. The last
+    layer is considered to have infinite thickness.
```
Comment (jcapriot) [inline geoana/kernels/tranverse_electric_reflections.py:50]:
```suggestion
    layer is considered to have infinite thickness. All physical properties 
    are defined starting from the top most layer. 
```

**Pull request merged**